Classes/Code/ScriptTagBuilder.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the "matomo_integration" extension for TYPO3 CMS.
+ *
+ * For the full copyright and license information, please read the
+ * LICENSE.txt file that was distributed with this source code.
+ */
+
+namespace Brotkrueml\MatomoIntegration\Code;
+
+use Brotkrueml\MatomoIntegration\Event\EnrichScriptTagEvent;
+use Psr\EventDispatcher\EventDispatcherInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+/**
+ * The tag builder creates the header script tag to include matomo
+ *
+ * @internal
+ */
+class ScriptTagBuilder
+{
+    private ServerRequestInterface $request;
+    private EventDispatcherInterface $eventDispatcher;
+
+    public function __construct(
+        EventDispatcherInterface $eventDispatcher
+    ) {
+        $this->eventDispatcher = $eventDispatcher;
+    }
+
+    /**
+     * Build the script start script tag with attributes
+     */
+    public function build(string $code): string
+    {
+        $attributes = $this->collectAttributes();
+
+        $attributes = array_map(static function (string $name, ?string $value = null): string {
+            if ($value === null) {
+                return $name;
+            }
+            if ($value === '') {
+                return $name;
+            }
+            return $name . '="' . htmlspecialchars($value) . '"';
+        }, array_keys($attributes), array_values($attributes));
+
+        $prepend = '';
+        if ($attributes !== []) {
+            $prepend = ' ';
+        }
+
+        return '<script' . $prepend . implode(' ', $attributes) . '>' . $code . '</script>';
+    }
+
+    /**
+     * Collect all attributes
+     *
+     * @return array<string,string|null>
+     */
+    private function collectAttributes(): array
+    {
+        $enrichScriptTagEvent = new EnrichScriptTagEvent($this->request ?? $this->getRequest());
+        /** @var EnrichScriptTagEvent $enrichScriptTagEvent */
+        $enrichScriptTagEvent = $this->eventDispatcher->dispatch($enrichScriptTagEvent);
+
+        $attributes = [];
+
+        if ($enrichScriptTagEvent->getId() !== '') {
+            $attributes['id'] = $enrichScriptTagEvent->getId();
+        }
+        if ($enrichScriptTagEvent->getType() !== '') {
+            $attributes['type'] = $enrichScriptTagEvent->getType();
+        }
+
+        foreach ($enrichScriptTagEvent->getDataAttributes() as $name => $value) {
+            $attributes['data-' . $name] = $value;
+        }
+
+        return $attributes;
+    }
+
+    public function setServerRequest(ServerRequestInterface $request): void
```

Comment:
Propose to move it up after the construct, so it is more clear it has to be initialised. Also name it setRequest() to be inline with core.

Reply:
Moved and renamed